### ISSUE TITLE
False positive in OptionalUnit rule

### ISSUE THREAD

**Issue by BraisGabin** [detekt/detekt#2452]
## Expected Behavior
```kotlin
val <T> T.exhaustive: T
    get() = this

enum class E { A, B }

fun a(value: E) {
    when (value) {
        E.A -> println("hi")
        E.B -> {
            if (System.currentTimeMillis() % 2 == 0L) {
                println("hi")
            }
            Unit
        }
    }.exhaustive
}
```

This code is correct. I need to add Unit at the end of `B` because `if` is not an expression so `exhaustive` fails.

## Observed Behavior
This code is flagged

## Context
I want to force exhaustive `when`s so I need to do things like this. Maybe I'm missing an easier and more elegant way to do it, if so, please tell me.

## Your Environment
* Version of detekt used: 1.6.0

**Comment by schalkms** [comment]
Unfortunately, the mentioned code snippet doesn't compile.

I don't know how you would want to detect this case in the `OptionalUnit` rule without having sophisticated data flow analysis.
Anyways, I think this is way too complex to handle.

**Comment by BraisGabin** [comment]
I editted it to a code that compiles.

I think that it's not that difficult... I'll give it a try.

**Comment by schalkms** [comment]
> I think that it's not that difficult... I'll give it a try.

I'm curious, how would you want to decide whether the `Unit` is unnecessary?
If you leave out the `.exhaustive` you don't need the `Unit` at the end.

**Comment by BraisGabin** [comment]
You are right, I wasn't thinking about that! I was just thinking about an `Unit` after an `if`... Yeap, it's a bit more difficult. I'll give it a try tomorrow. I can't leave my house so I have A LOT of time x"D

**Comment by schalkms** [comment]
I don’t think it’s a good idea to reimplement this just for the sake of fixing this false-positive. This could fix the original problem, but it results in other false-positives.
I’d suggest to take a look at the IntelliJ or Kotlin inspection that triggers this and maybe call the corresponding function.
Otherwise, we would add complexity that isn’t worth it.
